src/broadcast/broadcast.rs

```diff
@@ -362,44 +362,55 @@ impl<N: NodeIdT> Broadcast<N> {
         }
         // Upon receiving 2f + 1 matching Ready(h) messages, send full
         // `Echo` message to every node who hasn't sent us a `CanDecode`
-        let our_value = self.echos.get(self.our_id()).unwrap();
         if self.count_readys(hash) == 2 * self.netinfo.num_faulty() + 1 {
             // `send_echo` function should take care of which nodes to send the message
-            step.extend(self.send_echo(our_value)?);
+            step.extend(self.send_echo_remaining()?);
         }
 
         Ok(step.join(self.compute_output(hash)?))
     }
 
-    /// Sends an `Echo` message and handles it. Does nothing if we are only an observer.
-    fn send_echo(&mut self, p: Proof<Vec<u8>>) -> Result<Step<N>> {
-       if !self.netinfo.is_validator() {
+    /// Sends `Echo` message to all left nodes.
+    fn send_echo_left(&mut self) -> Result<Step<N>> {
+        if !self.netinfo.is_validator() {
             return Ok(Step::default());
         }
+        // Caller should make sure message is inserted. unwrap should never panic.
+        let p = self.echos.get(self.our_id()).unwrap().clone();
         let echo_msg = Message::Echo(p.clone());
         let mut step = Step::default();
-        // TODO: iterator not correctly describing right of our_id. Use cycle method on
+        // TODO: iterator not correctly describing left of our_id. Use cycle method on
         // iterator to correctly specify.
-        let right = self.netinfo.all_ids().skip(self.netinfo.num_correct()+self.pessimism_factor);
+        let left = self.netinfo.all_ids_except(self.proposer_id()).take(self.netinfo.num_correct()+self.pessimism_factor);
+        for id in left {
+            let msg = Target::Node(id.clone()).message(echo_msg.clone());
+            step.messages.push(msg);
+        }
+        let our_id = &self.our_id().clone();
+        Ok(step.join(self.handle_echo(our_id, p)?))
+    }
 
-        // Send `Echo` to all left nodes.
-        if !self.echo_sent_partial {
-            self.echo_sent_partial = true;
-            for id in right {
+    /// Sends `Echo` message to remaining nodes who haven't sent `CanDecode`
+    fn send_echo_remaining(&mut self) -> Result<Step<N>> {
+        // No need for echo_sent_partial check
+        self.echo_sent_full = true;
+        if !self.netinfo.is_validator() {
+            return Ok(Step::default());
+        }
+        // Caller should make sure message is inserted. unwrap should never panic.
+        let p = self.echos.get(self.our_id()).unwrap().clone();
```

Comment:
That's not true: The proposer might not have sent us a value (maliciously) or the message might be delayed. I _think_ it's correct to just return here if we don't have a value, i.e. if that `echos` entry doesn't exist: The fact that we arrived here means that the algorithm is going to be able to terminate even without our own `Echo` anyway, doesn't it?

Reply:
Yeah we should be able to terminate since we wouldn't have reached this function if we don't have atleast `N-2f` shards. Returning `Step::default()` should be okay I think.